src/lib/y2storage/boot_requirements_strategies/legacy.rb

```diff
@@ -26,48 +26,52 @@ module Y2Storage
   module BootRequirementsStrategies
     # Strategy to calculate the boot requirements in a legacy system (x86 without EFI)
     class Legacy < Base
-      GRUB_SIZE = DiskSize.KiB(256)
-      GRUBENV_SIZE = DiskSize.KiB(1)
-
       def initialize(*args)
         super
         textdomain "storage"
       end
 
       # @see Base#needed_partitions
       def needed_partitions(target)
-        raise Error if grub_in_mbr? && mbr_gap && !valid_mbr_gap?
         planned_partitions = super
         planned_partitions << grub_partition(target) if grub_partition_needed? && grub_partition_missing?
         planned_partitions
       end
 
       # Boot warnings in the current setup
       #
+      # Note that all are just warnings and we don't trigger real errors so far.
+      #
       # @return [Array<SetupError>]
       def warnings
         res = super
 
         if boot_ptable_type?(:gpt)
           res.concat(errors_on_gpt)
-        else
+        elsif boot_ptable_type?(:msdos)
           res.concat(errors_on_msdos)
+        else
+          res.concat(errors_on_plain_disk)
         end
 
         res
       end
 
     protected
 
-      # Whether the MBR gap is big enough
+      # Whether a BIOS GRUB partition will be needed.
       #
-      # @return [Boolean] true if the MBR gap is enough; false otherwise.
-      def valid_mbr_gap?
-        mbr_gap && mbr_gap >= GRUB_SIZE
-      end
-
+      # Note: this is a bit tricky.
+      #
+      # This function is intended to be used in #needed_partitions while
+      # creating a partition proposal. The catch is that when there is no
+      # partition table (yet) it is implicitly assumed that there will be a
+      # gpt created finally - so a grub partition is needed also in this
+      # case.
+      #
+      # @return [Boolean]
       def grub_partition_needed?
-        boot_ptable_type?(:gpt) && grub_part_needed_in_gpt?
+        (boot_ptable_type?(:gpt) || boot_ptable_type?(nil)) && grub_part_needed_in_gpt?
```

Comment:
This should not directly assume the default type will be GPT. This should rely on `#preferred_ptable_type`. For example, it's quite likely that for Raspberry Pi  we will have to modify `#preferred_ptable_type` to return msdos (looks like Raspi doesn't support GPT in the booting device). Or any other circumstances, that's what we have `#preferred_ptable_type` for.

That is, something like this (just from the top of my head):
```ruby
gpt_used =
  if boot_ptable_type?(nil)
    boot_disk.preferred_ptable_type.is?(:gpt)
  else
    boot_ptable_type?(:gpt)
  end
gpt_used && grub_part_needed_in_gpt?
```

Although is probably nicer to simply have a flag in `#boot_ptable_type?` to tell it to keep the old behavior. So we could have something like

```ruby
boot_ptable_type?(:gpt, fallback_to_preferred_type: true) && grub_part_needed_in_gpt?
```

(ok, the argument could have a shorter name).

Reply:
`boot_ptable_type` has cost me a lot of new gray hair already. And that even though it's only used withing the boot requirements code. I will try the first version if only to avoid the double semantics in `boot_ptable_type`.